File: 876-hand-of-straights/hand-of-straights.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupsize) {
        // store the total number of cards.
        int n = hand.size();
        // total number of cards must be divisible by groupsize so that every card can belong to exactly one group.
        if (n % groupsize != 0) {
            return false;
        }
        // store the frequency of every card.
        map<int, int> mp;
        // count how many times each card appears.
        for (int handnumber : hand) {
            mp[handnumber]++;
        }
        // continue creating groups until all cards have been used.
        while (!mp.empty()) {
            // always start the group with the smallest card that is still available.
            int current = mp.begin()->first;
            // try to create a consecutive group.
            for (int i = 0; i < groupsize; i++) {
                // calculate the card needed at the current position.
                int card = current + i;
                // if the required card is not available a consecutive group cannot be formed.
                if (mp.find(card) == mp.end()) {
                    return false;
                }
                // use one occurrence of this card.
                mp[card]--;
                // if all copies of this card have been used remove it from the frequency map.
                if (mp[card] < 1) {
                    mp.erase(card);
                }
            }
        }
        // every card was successfully used in consecutive groups.
        return true;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)

File: 876-hand-of-straights/hand-of-straights.py
# Brute Force Code & Optimal Code
class Solution:
    def isNStraightHand(self, hand: List[int], groupsize: int) -> bool:
        # store the total number of cards
        n = len(hand)
        # total number of cards must be divisible by groupsize so that every card can belong to exactly one group
        if n % groupsize != 0:
            return False
        # store the frequency of every card
        mp = {}
        # count how many times each card appears
        for handnumber in hand:
            mp[handnumber] = (mp.get(handnumber, 0) + 1)
        # continue creating groups until all cards have been used
        while mp:
            # always start groups with the smallest card that is still avaiilable 
            current = min(mp)
            # try to create consecutive group
            for i in range(groupsize):
                # calculate the card needed at the current position
                card = current + i
                # if the required card is not avaiilable consecutive group cannot be formed
                if mp.get(card, 0) == 0:
                    return False
                # use one occurence of this card
                mp[card] -= 1
                # if all copies of this card have been used remove if form the frequency map
                if mp[card] < 1:
                    del mp[card]
        # every card was sucessfully used in consecutive groups
        return True

# Time Complexity : O(N)
# Space Complexity : O(N)